### PR TITLE
feat(#352): remove compasflow and extref bindings

### PR DESCRIPTION
**Pull request by AliouDIAITE**

### CODE REVIEW

**Review by samirromdhani — APPROVED** [review]

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ExtRefEditorService.java [hunk 1]
```diff
@@ -399,4 +413,40 @@ private String computeDaiValue(AbstractLNAdapter<?> lnAdapter, TExtRef extRef, S
...
 
+    @Override
+    public void debindCompasFlowsAndExtRefsBasedOnVoltageLevel(SCL scd) {
+        LdeviceService ldeviceService = new LdeviceService();
+        scd.getSubstation()
+                .stream()
+                .flatMap(tSubstation -> tSubstation.getVoltageLevel().stream())
+                .filter(tVoltageLevel -> !"0".equals(tVoltageLevel.getName()))
+                .forEach(tVoltageLevel -> {
+                    String flowSource = voltageCodification.get(tVoltageLevel.getName());
```
Comment (gleizesDor) [inline sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ExtRefEditorService.java:424]: put this variable closer to the place it's used

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ExtRefEditorService.java [hunk 2]
```diff
@@ -399,4 +413,40 @@ private String computeDaiValue(AbstractLNAdapter<?> lnAdapter, TExtRef extRef, S
...
+                    scd.getIED().stream()
+                            .flatMap(ldeviceService::getLdevices)
+                            .forEach(tlDevice -> {
```
Comment (gleizesDor) [inline sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ExtRefEditorService.java:427]: you can put tCompasFlow.isSetFlowSourceVoltageLevel() and isSetExtRefiedName in a filter condition, as it is used in both case : 
```java
PrivateUtils.getPrivateStream(tInputs.getPrivate(), TCompasFlow.class)
                                    .filter(TCompasFlow::isSetFlowSourceVoltageLevel)
                                    .filter(TCompasFlow::isSetExtRefiedName)
                                    .forEach(tCompasFlow -> {
```

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ExtRefEditorService.java [hunk 3]
```diff
@@ -399,4 +413,40 @@ private String computeDaiValue(AbstractLNAdapter<?> lnAdapter, TExtRef extRef, S
...
+                                TInputs tInputs = tlDevice.getLN0().getInputs();
+                                List<TCompasFlow> compasFlows = PrivateUtils.getPrivateStream(tInputs.getPrivate(), TCompasFlow.class).toList();
+                                compasFlows.forEach(tCompasFlow -> {
+                                    if (flowSource == null && tCompasFlow.isSetFlowSourceVoltageLevel() && tCompasFlow.isSetExtRefiedName()) {
+                                        //debind all compas flow
+                                        extRefService.clearCompasFlowBinding(tCompasFlow);
+                                    }
+                                    if (flowSource != null && tCompasFlow.isSetExtRefiedName() && tCompasFlow.isSetFlowSourceVoltageLevel() && !tCompasFlow.getFlowSourceVoltageLevel().equals(flowSource)) {
```
Comment (gleizesDor) [inline sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ExtRefEditorService.java:435]: I think I prefer "else-if" method : 
```java
String flowSource = voltageCodification.get(voltageLevel);
                                        if (flowSource == null) {
                                            //debind all compas flow
                                            extRefService.clearCompasFlowBinding(tCompasFlow);
                                        } else if (!tCompasFlow.getFlowSourceVoltageLevel().equals(flowSource)) {
                                            //debind extRef
``

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ExtRefEditorService.java [hunk 4]
```diff
@@ -399,4 +413,40 @@ private String computeDaiValue(AbstractLNAdapter<?> lnAdapter, TExtRef extRef, S
...
+                .stream()
+                .flatMap(tSubstation -> tSubstation.getVoltageLevel().stream())
+                .filter(tVoltageLevel -> !"0".equals(tVoltageLevel.getName()))
+                .forEach(tVoltageLevel -> {
+                    String flowSource = voltageCodification.get(tVoltageLevel.getName());
+                    scd.getIED().stream()
+                            .flatMap(ldeviceService::getLdevices)
+                            .forEach(tlDevice -> {
+                                TInputs tInputs = tlDevice.getLN0().getInputs();
+                                List<TCompasFlow> compasFlows = PrivateUtils.getPrivateStream(tInputs.getPrivate(), TCompasFlow.class).toList();
```
[the remark this answers is not in the document]
Comment (AliouDIAITE) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ExtRefEditorService.java:429]: Ok, good .I forgot it, it was for debugg purpose,

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ExtRefService.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    /**
+     * Find CompasFlows that match given ExtRef
+     *
+     * @param inputs  inputs containing Privates CompasFlow and TExtRefs
+     * @param tExtRef extRef to match
+     * @return list of matching CompasFlows
+     */
+    public Stream<TCompasFlow> getMatchingCompasFlows(TInputs inputs, TExtRef tExtRef) {
```
[the remark this answers is not in the document]
Comment (AliouDIAITE) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ExtRefService.java:38]: Here the goal is to fetch all CompasFlows corresponding to given TExtRef. I think it's ok

[the remark this answers is not in the document]
Comment (samirromdhani) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ExtRefService.java:38]: Ok

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ExtRefService.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    /**
+     * Retrieves ExtRefs corresponding to given
+     *
+     * @param inputs      node containing CompasFlows and ExtRefs
+     * @param tCompasFlow for which Extrefs are searched
+     * @return stream of matching ExtRefs
+     */
+    public Stream<TExtRef> getMatchingTextRef(TInputs inputs, TCompasFlow tCompasFlow) {
```
[the remark this answers is not in the document]
Comment (samirromdhani) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ExtRefService.java:50]: OK
```suggestion
    public Stream<TExtRef> getMatchingExtRef(TInputs inputs, TCompasFlow tCompasFlow) {
```

[the remark this answers is not in the document]
Comment (AliouDIAITE) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ExtRefService.java:50]: ok

**Pull request merged**